client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -150,20 +150,17 @@ private PositionInternal getPosition() {
     @GuardedBy("readers")
     private String updateGroupStateIfNeeded() throws ReinitializationRequiredException {
         try {
-            String checkpoint = groupState.getCheckpoint();
-            if (checkpoint == null) {
-                if (atCheckpoint) {
-                    releaseSegmentsIfNeeded();
-                    atCheckpoint = false;
-                }
-                acquireSegmentsIfNeeded();
-                return null;
-            } else {
-                log.info("{} at checkpoint {}", this, checkpoint);
-                groupState.checkpoint(checkpoint, getPosition());
```

Comment:
It would be great in issue #2671 to see a definition of what is considered as a _received_ and _processed_ checkpoint, as well as the expected behavior of the reader upon these situations. With my limited understanding, from the viewpoint of a client a checkpoint is _processed_ as soon as its positions are persisted in the `ReaderGroupState`, which is actually what the `else` block was doing in the previous code version. This was executed as soon as an ongoing checkpoint was _received_ (`getCheckpoint()`), which seems natural. Now, the analogous piece of code in the new version does nothing when an ongoing checkpoint is _received_ and offloads to the next `readNextEvent()` call the task of updating its positions in the `ReaderGroupState` (i.e., _processing_). What are the additional guarantees that this delay is providing to the application (considering multiple readers)? As a side effect, if `StreamCuts` are built upon persisted reader positions, why do not update them as soon as possible? Any clarification regarding these questions will be appreciated.

Reply:
Sorry for the confusion: In the issue the Client refers to the client application not the client library. 
The code, without this change, is updating the reader group state object and hence potentially unblocking the future returned from initiateCheckpoint() before actually returning the checkpoint to the client application. This change prevents that. It ensures that the application has done whatever it was going to do with that object and asked for the next event before allowing the future to complete.

Reply:
I see, so the desired behavior is that the `CompletableFuture<Checkpoint>` in `initiateCheckpoint()` returns _strictly after_ all the application clients have executed the logic associated with the checkpoint event. I think that if there is no comment in the `initiateCheckpoint()` method Javadoc already describing this, it would be necessary to add one.

Reply:
Let's imagine a user wrote an application that, as part of the logic related to reading a checkpoint event, executes a `checkpoint.join()` on the result of a previous `initiateCheckpoint()` call. Would it be possible that `checkpoint.join()` returns with the current version of the code, but gets blocked after this PR? If this hypothesis is not wrong, could this be interpreted as a "breaking" change?

Reply:
It will be a change, but it is not breaking because the future will still complete with the same information. The only difference is timing. 

I updated the Javadoc for initiateCheckpoint()

Reply:
Thanks for considering updating the Javadoc. You are right, the only change is regarding the timing when the `CompletableFuture<Checkpoint>` is returned; I referred to the term "breaking change" in the sense that there could be applications today working with Pravega that read events and rely on joining a `CompletableFuture<Checkpoint>` before executing the next `readNextEvent()`, and perhaps, due to the behavioral change of the completion of the future now they may get blocked. Do you think that this situation is possible?